[quant][graphmode] Refactor bias and weight check and add aten::linear pattern

jerryzh168 opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* #30476 [quant][graphmode] Add tests for quantizing traced models
* #30475 Refactor test_quantization.py and enable `test_nested`
* **#30474 [quant][graphmode] Refactor bias and weight check and add aten::linear pattern**

Summary:
There are some common parts in `isBiasOfConvOrLinear` and `isWeightOfConvOrLinear`, we can factor
them out, the refactor will allow for easier extension of new patterns

Test Plan:
python test/test_jit.py
python test/test_quantization.py

Reviewers:
mvz

Subscribers:

Tasks:

Tags:

Differential Revision: [D18795725](https://our.internmc.facebook.com/intern/diff/D18795725)

jerryzh168 committed:
[quant][graphmode] Refactor bias and weight check and add aten::linear pattern

Summary:
There are some common parts in `isBiasOfConvOrLinear` and `isWeightOfConvOrLinear`, we can factor
them out, the refactor will allow for easier extension of new patterns

Test Plan:
python test/test_jit.py
python test/test_quantization.py

Reviewers:
mvz

Subscribers:

Tasks:

Tags:

[ghstack-poisoned]

lly-zero-one commented on torch/csrc/jit/passes/quantization.cpp:
```diff
@@ -202,35 +202,61 @@ graph(%input, %weight, %bias, %4):
-    }
-  }
-  return false;
+  return matchArgPattern(v, {{"conv2d", 1}, {"linear", 1}}, {{"linear", 2}});
```
Why not directly call the matchArgPattern function and give some explanation comments?

jerryzh168 replied:
I feel structuring them in functions is clearer than calling lower level functions directly

ZolotukhinM commented on torch/csrc/jit/passes/quantization.cpp:
```diff
@@ -202,35 +202,61 @@ graph(%input, %weight, %bias, %4):
 }
 
+bool isBiasOfConvOrLinear(Value* v) {
+  return matchArgPattern(v, {{"conv2d", 2}, {"linear", 2}}, {{"linear", 3}});
```
This is concerning, because for `{"linear", 2}` we would say that it is both bias and weight. That's probably incorrect. I think it might work correctly now as a side effect of how we invoke these two functions, but this should not be that fragile.

jerryzh168 replied:
first vector is for aten function call, second is for `CallFunction` call, so although {"linear", 2} appears twice, it mean different things

jamesr66a replied:
@jerryzh168 Can we make this more explicit at the callsites? The separation with the braces is pretty subtle here and I'm not surprised it's confusing.

Ideally we would have something like python kwargs. I think we can emulate this in C++ with something like:

```
using AtenFuncArgs = std::vector<FuncArg>;
using CallFuncArgs = std::vector<FuncArg>;
bool matchArgPattern(
    Value* v,
    const AtenFuncArgs& aten_func_args,
    const CallFuncArgs& call_func_args) {
```

and then

```
  return matchArgPattern(v, AtenFuncArgs{{"conv2d", 2}, {"linear", 2}}, CallFuncArgs{{"linear", 3}});
```

Or something liek that

jerryzh168 committed:
Update on "[quant][graphmode] Refactor bias and weight check and add aten::linear pattern"


Summary:
There are some common parts in `isBiasOfConvOrLinear` and `isWeightOfConvOrLinear`, we can factor
them out, the refactor will allow for easier extension of new patterns

Test Plan:
python test/test_jit.py
python test/test_quantization.py

Reviewers:
mvz

Subscribers:

Tasks:

Tags:

[ghstack-poisoned]

jerryzh168 committed:
Update on "[quant][graphmode] Refactor bias and weight check and add aten::linear pattern"


Summary:
There are some common parts in `isBiasOfConvOrLinear` and `isWeightOfConvOrLinear`, we can factor
them out, the refactor will allow for easier extension of new patterns

Test Plan:
python test/test_jit.py
python test/test_quantization.py

Reviewers:
mvz

Subscribers:

Tasks:

Tags:

[ghstack-poisoned]

jamesr66a reviewed (approved):
This looks good overall, and should improve readability. Just had one small comment to potentially further improve readability.

jerryzh168 committed:
Update on "[quant][graphmode] Refactor bias and weight check and add aten::linear pattern"


Summary:
There are some common parts in `isBiasOfConvOrLinear` and `isWeightOfConvOrLinear`, we can factor
them out, the refactor will allow for easier extension of new patterns

Test Plan:
python test/test_jit.py
python test/test_quantization.py

Reviewers:
mvz

Subscribers:

Tasks:

Tags:

Differential Revision: [D18795725](https://our.internmc.facebook.com/intern/diff/D18795725)

[ghstack-poisoned]

jerryzh168 committed:
Update on "[quant][graphmode] Refactor bias and weight check and add aten::linear pattern"


Summary:
There are some common parts in `isBiasOfConvOrLinear` and `isWeightOfConvOrLinear`, we can factor
them out, the refactor will allow for easier extension of new patterns

Test Plan:
python test/test_jit.py
python test/test_quantization.py

Reviewers:
mvz

Subscribers:

Tasks:

Tags:

Differential Revision: [D18795725](https://our.internmc.facebook.com/intern/diff/D18795725)

[ghstack-poisoned]
